### PR TITLE
L2 -> L1 Connectivity & Message Passing

### PR DESCRIPTION
**Pull request by willmeister**
## Description
- Adds L1 node configuration and default deployment if connection url not specified
- Adds L2 -> L1 connectivity by searching L2 events 
- Adding docker / environment var configuration for L1 and L2 nodes

## Questions
- Any reason to queue (enforce order for) L2 -> L1 messages or should we just fire them as soon as we have them?

### Fixes
- Fixes # [YAS-236](https://optimists.atlassian.net/browse/YAS-236)

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/rollup-full-node/src/app/message-submitter.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+export class DefaultL2ToL1MessageSubmitter implements L2ToL1MessageSubmitter {
+  public static readonly LAST_CONFIRMED_KEY: Buffer = Buffer.from(
+    'last_confirmed'
+  )
+  public static readonly LAST_SUBMITTED_KEY: Buffer = Buffer.from(
+    'last_submitted'
+  )
+  public static readonly LAST_QUEUED_KEY: Buffer = Buffer.from('last_queued')
+
+  private lastSubmitted: BigNumber
```
Comment (masonforest) [inline packages/rollup-full-node/src/app/message-submitter.ts]: Are we sure we need to use `BigNumber`s here? From a quick calculation we could run a million transactions a second for over 200 years before overflowing a regular JavaScript number: `Number.MAX_SAFE_INTEGER/(60*60*24*365*1000000) => 285.61641472415624` or is there another benefit of BigNumbers that I'm not seeing?

File: packages/rollup-full-node/src/app/message-submitter.ts [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+  public static getMessageKey(messageNumber: BigNumber): Buffer {
+    return Buffer.from(`MESSAGE_${add0x(messageNumber.toString('hex'))}`)
+  }
+
+  /***********
+   * GETTERS *
+   ***********/
+  public getLastSubmitted(): BigNumber {
```
Comment (masonforest) [inline packages/rollup-full-node/src/app/message-submitter.ts]: Any interest in using [TypeScript Accessors](https://www.typescriptlang.org/docs/handbook/classes.html#accessors)?
```suggestion
  public get lastSubmitted(): BigNumber {
```
I'm also open to the idea of dropping getters completely and just accessing state directly 😬 That's probably a little to `YOLO` though.

File: packages/rollup-core/src/app/serialization/abi-parsers.ts [hunk 3]
```diff
@@ -14,3 +15,16 @@ export const abiDecodeRollupBlock = (abiEncoded: string): RollupBlock => {
     transactions: [],
   }
 }
+
+export const abiDecodeL2ToL1Message = (abiEncoded: string): L2ToL1Message => {
+  const [nonce, ovmSender, callData] = abi.decode(
+    ['address', 'bytes'],
```
[the remark this answers is not in the document]
Comment (willmeister) [reply packages/rollup-core/src/app/serialization/abi-parsers.ts]: Probably an error 😄

File: packages/rollup-full-node/src/app/utils/local-l1-node.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    mnemonic,
+  })
+
+  const wallet = getWallets(provider)[0]
+  await deployL2ToL1MessagePasser(wallet)
+
+  return provider
+}
+
+const deployL2ToL1MessagePasser = async (wallet: Wallet): Promise<Contract> => {
```
Comment (ben-chain) [inline packages/rollup-full-node/src/app/utils/local-l1-node.ts]: I think this should be named `deployL2ToL1MessageReciever`

Comment (willmeister) [reply packages/rollup-full-node/src/app/utils/local-l1-node.ts]: 100% Thanks!

**Comment by ben-chain** [comment]
Oh also, WRT the question here about ordering -- I think fine to ignore ordering here.  Ordering will be more strongly enforced when we incorporate L2->L1 messages into the rollup state itself, and I can't really see many use cases where it would be super important anyway.

File: docker-compose.yml [hunk 5]
```diff
@@ -18,17 +19,29 @@ services:
...
+       - geth-l2-data:/mnt/geth/l2
     networks:
       - internal
     build:
       context: docker/geth
       dockerfile: Dockerfile
     environment:
-      VOLUME_PATH: /mnt/docker
+      VOLUME_PATH: /mnt/geth/l2
+  geth_l1:
```
[the remark this answers is not in the document]
Comment (willmeister) [reply docker-compose.yml]: Talked through this with you, but just documenting it here. In the short run, local seems to make sense, but in the long run, you're totally right.

**Pull request merged**